Added fishing minigame; Still bugs to fix

fullspeccoder opened the pull request:
## Fishing Minigame

Had to work through some major changes due to major refactor, but managed to figure it out at least for now. I implemented the fishing minigame for the project - but it doesn't work on mobile just yet. I was thinking of maybe implementing that version where the user stays in a spot and has to tap to get a fish. 

***Implementation***
---

A game that allows you to cast a hook and will catch fish on the way up. In this version, you can move side to side (albeit slowly because that makes it more challenging). The fishes spawn in random places between boundaries with random sprites used from the fish.png image from the sprites folder. I used a strategy of loading assets before the game starts, configuring the assets to be used correctly, then allow the user to play the game.

***Things To Improve***
---

Upon successful catches once you leave the minigame, you should be able to keep the fish you caught. I'm not sure what this could look like; whether that means you can keep the specific names of the fish, or keeping the weights. There are so many possibilities!

Will work out details on getting state to work versus the triggerFishingAnimations and triggerNormalAnimations functions. 

Hopefully the structure for the minigame won't affect the rest of the code in the project, I tried to modularize it as much as possible away from the rest of the code.

fullspeccoder commented:
Through the code changes I realized that all I really should have done was pass in a type of fishing scene then create the string scene so that the player could go back to the map they were on.

r4pt0s commented:
@fullspeccoder that is what I meant on Discord. I tried to reduce the boilerplate code for scene addition. Hopefully it will help you to add your part more easily

fullspeccoder commented:
I finally added the minigame! In a similar style to what you've refactored it to @r4pt0s, I have also fixed the scaling but I see that it was also fixed in another PR as well. Fixing the scaling helped with an overall issue I was having with the game. I tested escaping out of the game and it worked okay, the game works as intended, and overall seems to work okay. Let me know if there's any changes!

r4pt0s commented:
Hi @fullspeccoder,

I looked and tested it.
There is a problem with the mobile view of your mini game.
![Screenshot_2024-10-12-17-33-45-82_40deb401b9ffe8e1df2f1cc5ba480b12](https://github.com/user-attachments/assets/431e3e38-7f07-41d4-96a4-bbc2fbc0ac22)

On desktop everything looks and works as expected. 

Also, could you please merge the current main branch into your PR 🙏🏻

r4pt0s reviewed (changes requested):
^

fullspeccoder commented:
Hello again! I have finally figured out a way to make the game look good on Mobile, as the screen was scrunched up at the top. Hopefully this is the absolute winner! Let me know if you need any more changes. @r4pt0s

r4pt0s commented on src/constants.js:
```diff
@@ -1,4 +1,13 @@
+const scaleFactorWidth = (canvas.offsetWidth - leftPanel.offsetWidth * 2) / 480;
+const scaleFactorHeight =
+    (canvas.offsetHeight - (header.offsetHeight + footer.offsetHeight)) / 320;
+export const scaleFactor = Math.min(scaleFactorWidth, scaleFactorHeight);
```
This doesn't work on desktop nor on mobile. Only a scale factor of 2 looks good what I figured out.
If we calculate the scale factor here, then everything will be to small globally.
Maybe you can set the scale factor for the mini game only once it started and switch back to the default one after exiting the fishing game?

r4pt0s commented on src/main.js:
```diff
@@ -24,12 +26,13 @@ k.scene('forest_junction', (enter_tag) =>
 k.scene('lose', loseScreen);
 
-k.go('start');
+k.go('forest', 'spawn_bottom');
```
can you revert please to start scene again?

fullspeccoder replied:
I will fix this issue once I have finally fixed the game. If you can, try to remind me lol

r4pt0s commented on src/scenes/fishing_minigame/factories/fish.factory.js:
can you move this file to the global factories directory under src?
Maybe we will reuse it somewhere else

fullspeccoder replied:
I have fixed this issue.

r4pt0s commented on src/scenes/fishing_minigame/factories/hook.factory.js:
can you move this file to the global factories directory under src?
Maybe we will reuse it somewhere else

fullspeccoder replied:
I have fixed this issue

r4pt0s commented on src/scenes/fishing_minigame/factories/player.factory.js:
can you not just use the default player factory to generate the character and just replace the sprite with the characters_fishing sprites? :thinking: 
Only in the case where the finishing minigame started and revert on exit the fishing game

fullspeccoder replied:
I have fixed this issue.

r4pt0s commented on src/scenes/fishing_minigame/init/map.init.js:
can you not just use the exiting map.init.js file?
If it is because of the props for the map, then you can just add them manually once you get the map object returned

fullspeccoder replied:
So I added a mapConfig object. It turns out this was the winner for that. I changed the appropriate lines for the mapConfig object. In my original init file, I had to ensure that there were game states for that specific scene, so I had to add additional properties to the map. This is why I created the mapConfig.

`
const mapConfig = {
  additionalTags = [],
  additionalProperties: {
    pressed: false,
    pressedTwice: false,
    fishSpawnerTimer: ...,
    score: 0,
  },
  mapOffset: ..., // Was premade before this object was made
}
`

This is in fishing.js under the minigame folder

r4pt0s replied:
Wooohhhhooooo, that's great 🎉

r4pt0s commented on src/main.js:
```diff
@@ -24,12 +26,13 @@ k.scene('forest_junction', (enter_tag) =>
 k.scene('lose', loseScreen);
 
-k.go('start');
+k.go('fishing');
```
Just revert this back to start and we are good to go 🙌🏻 @fullspeccoder

fullspeccoder replied:
Done! Thank you so much for the thoughts and critique on the game 😌 I'm glad it'll finally be in the game!

r4pt0s reviewed (approved):
🔥🔥🔥
